src/jetstream/pull_subscription.rs

```diff
@@ -0,0 +1,369 @@
+    ///
+    /// let message = consumer.next_timeout(Duration::from_millis(1000))?;
+    /// println!("Received message: {:?}", message);
+    ///
+    /// // timeout on second, as there are no messages.
+    /// let message = consumer.next_timeout(Duration::from_millis(100));
+    /// assert!(message.is_err());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn next_timeout(&self, timeout: Duration) -> io::Result<Message> {
+        match self.0.messages.recv_timeout(timeout) {
+            Ok(message) => Ok(message),
+            Err(channel::RecvTimeoutError::Timeout) => Err(io::Error::new(
+                io::ErrorKind::TimedOut,
+                "next_timeout: timed out",
+            )),
+            Err(channel::RecvTimeoutError::Disconnected) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "next_timeout: unsubscribed",
+            )),
+        }
+    }
+
+    /// Returns blocking Subscription Iterator containing messages.
+    /// Iterator yields `None` when whole batch is processed.
+    pub fn fetch<I: IntoFetchOptions>(&self, batch: I) -> io::Result<BatchIter<'_>> {
+        let batch_options = batch.into_fetch_opts();
+        self.request_batch(batch_options)?;
+        Ok(self.batch_iter(batch_options.batch))
+    }
+
+    pub fn try_fetch(&self, batch: usize) -> io::Result<TryBatchIter<'_>> {
+        self.request_batch(batch)?;
+        Ok(self.try_batch_iter(batch))
+    }
+
+    pub fn timeout_fetch<I: IntoFetchOptions>(
+        &self,
+        batch: I,
+        timeout: Duration,
+    ) -> io::Result<TimeoutBatchIter<'_>> {
+        let batch_options = batch.into_fetch_opts();
+        self.request_batch(batch_options)?;
+        Ok(self.timeout_batch_iter(batch_options.batch, timeout))
+    }
+
+    pub fn request_batch<I: IntoFetchOptions>(&self, batch: I) -> io::Result<()> {
+        let batch_opts = batch.into_fetch_opts();
+
+        let subject = format!(
+            "{}CONSUMER.MSG.NEXT.{}.{}",
+            self.0.context.api_prefix(),
+            self.0.info.stream_name,
+            self.0.info.name,
+        );
+
+        let request = serde_json::to_vec(&batch_opts)?;
+
+        self.0.context.connection.publish_with_reply_or_headers(
+            &subject,
+            Some(self.0.inbox.as_str()),
+            None,
+            request,
+        )?;
+        Ok(())
+    }
+
+    pub(crate) fn batch_iter(&self, batch: usize) -> BatchIter<'_> {
+        BatchIter {
+            batch_size: batch,
+            processed: 0,
+            subscription: self,
+        }
+    }
+
+    pub(crate) fn try_batch_iter(&self, batch: usize) -> TryBatchIter<'_> {
+        TryBatchIter {
+            batch_size: batch,
+            processed: 0,
+            subscription: self,
+        }
+    }
+    pub(crate) fn timeout_batch_iter(
+        &self,
+        batch: usize,
+        timeout: Duration,
+    ) -> TimeoutBatchIter<'_> {
+        TimeoutBatchIter {
```

Comment:
Nit: a bit excessive with the helpers maybe? seems they're only called once anyway so inline imo but, your call.